### PR TITLE
Paper reviewing email notifications

### PR DESCRIPTION
**Pull request by nop33**
Preview of the emails (text might have changed): http://imgur.com/a/gulmF

### CODE REVIEW

File: indico/modules/events/papers/templates/emails/base.html [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+{% block footer %}{% endblock %}
+
+{% block subject -%}
+    {% block subject_message %}Call for papers{% endblock %}
+    {%- block subject_name %}{% endblock %}
+{%- endblock %}
+
+{% block body %}
+    <div class="mail" style="width: 600px; font-family: Arial; border: 1px solid #dfdfdf; background-color: #fcfcfc; margin: auto auto 40px auto;">
+        <div class="header" style="background-color: #5D95EA; color: #eee; padding: 10px 20px;">
```
Comment (mvidalgarcia) [inline indico/modules/events/papers/templates/emails/base.html]: Hexadecimal notation should be lowercase according to our set of sass-lint rules.

File: indico/modules/events/papers/templates/emails/base.html [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+            <h3 style="margin: 0; margin-top: 10px; padding-top: 10px; border-top: 1px solid #eee;">
+                {% block header_subtitle %}{% endblock %}
+                <span style="float: right; font-style: italic;">
+                    {% block header_subtitle_ref %}{% endblock %}
+                </span>
+            </h3>
+        </div>
+        <div class="content" style="padding: 10px 20px; color: #555;">
+            {% block content %}
+                <div style="border-bottom: 1px solid #EBEBEB;">
```
Comment (mvidalgarcia) [inline indico/modules/events/papers/templates/emails/base.html]: Same here.

File: indico/modules/events/papers/templates/emails/base.html [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                    {% block content_header %}{% endblock %}
+                </div>
+                {% block content_body %}{% endblock %}
+                <div>
+                    {% block content_footer %}{% endblock %}
+                </div>
+            {% endblock %}
+        </div>
+        <div style="text-align: center; padding: 10px; background: #EBEBEB; font-size: small; color: #999;">
```
Comment (mvidalgarcia) [inline indico/modules/events/papers/templates/emails/base.html]: And here.

File: indico/modules/events/papers/templates/emails/revision_submission_to_judge.html [hunk 4]
```diff
@@ -0,0 +1,12 @@
+{% extends 'events/papers/emails/revision_submission_base.html' %}
+
+{% block content %}
+    <p>Dear {{ receiver.first_name }},</p>
+    <p>An author has submitted a new revision for the paper <strong>"{{ revision.paper.contribution.title }}"</strong>.</p>
```
Comment (ThiefMaster) [inline indico/modules/events/papers/templates/emails/revision_submission_to_judge.html]: This is the same as in the reviewer email. Can we use a base class to avoid  repeating it? All that changes is the next paragraph.

File: indico/modules/events/papers/notifications.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    roles_to_notify = paper_reviewing_settings.get(event, 'notify_on_paper_submission')
+    if PaperReviewingRole.judge in roles_to_notify:
+        for judge in revision.paper.contribution.paper_judges:
+            template = get_template_module('events/papers/emails/revision_submission_to_judge.html', event=event,
+                                           revision=revision, receiver=judge)
+            email = make_email(to_list=judge.email, template=template, html=True)
+            send_email(email, event=event, module='Papers', user=session.user)
+    reviewers = set()
+    if PaperReviewingRole.layout_reviewer in roles_to_notify:
+        reviewers = revision.paper.cfp.layout_reviewers
```
Comment (ThiefMaster) [inline indico/modules/events/papers/notifications.py]: I'd use `|=` for consistency. That way you also guarantee  that the next `|=` won't change anything (imagine if the `layout_reviewers` property returned something that's not generated on the fly - then the in-place change 2 lines below could end up modifying data in the DB)

File: indico/modules/events/papers/operations.py [hunk 6]
```diff
@@ -302,6 +321,15 @@ def create_comment(paper, text, visibility, user):
     comment = PaperReviewComment(user=user, text=text, visibility=visibility)
     paper.last_revision.comments.append(comment)
     db.session.flush()
+    receivers = {x for x in paper.cfp.judges}
```
Comment (ThiefMaster) [inline indico/modules/events/papers/operations.py]: Why email judges not assigned to the paper? The might not even be able to see that paper.

File: indico/modules/events/papers/operations.py [hunk 7]
```diff
@@ -302,6 +321,15 @@ def create_comment(paper, text, visibility, user):
...
+    if visibility == PaperCommentVisibility.contributors or visibility == PaperCommentVisibility.reviewers:
+        receivers |= paper.cfp.layout_reviewers if paper.cfp.layout_reviewing_enabled else {}
```
Comment (ThiefMaster) [inline indico/modules/events/papers/operations.py]: same here

File: indico/modules/events/papers/operations.py [hunk 8]
```diff
@@ -302,6 +321,15 @@ def create_comment(paper, text, visibility, user):
...
+        receivers |= paper.cfp.content_reviewers if paper.cfp.content_reviewing_enabled else {}
+    if visibility == PaperCommentVisibility.contributors:
+        receivers |= {x.person.user for x in paper.contribution.person_links}
```
Comment (ThiefMaster) [inline indico/modules/events/papers/operations.py]: a person may not have an indico user associated so you should check for this - but I think it'd actually be better to use the person.email (which may also be empty, so take care of that!) here.

File: indico/modules/events/papers/templates/emails/added_to_reviewing_team.html [hunk 9]
```diff
@@ -0,0 +1,25 @@
+{% extends 'events/papers/emails/base.html' %}
+
+{% block subject_message %}Added in paper reviewing team{% endblock %}
+
```
Comment (ThiefMaster) [inline indico/modules/events/papers/templates/emails/added_to_reviewing_team.html]: Doesn't it look less strange without empty lines between one-line blocks? Probably just a matter of personal taste though..

File: indico/modules/events/papers/templates/emails/added_to_reviewing_team.html [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+{% block header_subtitle %}Added in paper reviewing team{% endblock %}
+
+{% block content %}
+    <p>Dear {{ receiver.first_name }},</p>
+    <p>You have been chosen as a <strong>{{ role.title }}</strong> for the conference <strong>{{ event.title }}</strong> in order to help with the paper reviewing process.</p>
+    {% if role.name == 'judge' %}
+        <p>Once there are some papers for you to judge, you will be able to do so in the judging area page:</p>
+        <p>
+            <a href="{{ url_for('papers.papers_list', event, _external=true) }}">
+                {{- url_for('papers.papers_list', event, _external=true) -}}
```
Comment (ThiefMaster) [inline indico/modules/events/papers/templates/emails/added_to_reviewing_team.html]: You need to add an explicit `management=false` here and in the other emails to make sure these emails always point to the page in the display area and not the equivalent in the management area.

Comment (ThiefMaster) [inline indico/modules/events/papers/templates/emails/added_to_reviewing_team.html]: FYI this applies to all the endpoints defined in the `for prefix, is_management in ...` loop, but I believe `papers_list` is the only one you are using.

File: indico/modules/events/papers/templates/emails/base.html [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+{% block footer %}{% endblock %}
+
+{% block subject -%}
+    {% block subject_message %}Call for papers{% endblock %}
+    {%- block subject_name %}{% endblock %}
+{%- endblock %}
+
+{% block body %}
+    <div class="mail" style="width: 600px; font-family: Arial; border: 1px solid #dfdfdf; background-color: #fcfcfc; margin: auto auto 40px auto;">
```
Comment (ThiefMaster) [inline indico/modules/events/papers/templates/emails/base.html]: What's the point of setting classes when we can't use non-inline CSS rules?
Also, do we really want to specify only `Arial` and not at least some fallback font available on non-Windows systems?

File: indico/modules/events/papers/templates/emails/judgment_reset_to_author.html [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+{% block subject_message %}Paper assessment withdrawn{% endblock %}
+
+{% block header_subtitle %}Paper assessment withdrawn{% endblock %}
+
+{% block header_subtitle_ref %}ref #{{ contribution.friendly_id }}{% endblock %}
+
+{% block content %}
+    <p>Dear {{ receiver.first_name }},</p>
+    <p>Your paper <strong>{{ contribution.title }}</strong> has been put back to review, after the assessment has been withdrawn.</p>
+    <p>To see your paper follow the link to your paper page:</p>
```
Comment (ThiefMaster) [inline indico/modules/events/papers/templates/emails/judgment_reset_to_author.html]: "your paper's timeline"?

File: indico/modules/events/papers/templates/emails/judgment_to_author.html [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+{% block subject_message %}Paper assessment{% endblock %}
+
+{% block header_subtitle %}Paper assessment{% endblock %}
+
+{% block header_subtitle_ref %}ref #{{ contribution.friendly_id }}{% endblock %}
+
+{% block content_header %}
+    <p>Dear {{ receiver.first_name }},</p>
+    {% if paper.state.name == 'accepted' %}
+        <p>Congratulations! A judge has <strong>accepted</strong> your paper <strong>{{ contribution.title }}</strong>.</p>
```
Comment (ThiefMaster) [inline indico/modules/events/papers/templates/emails/judgment_to_author.html]: Isn't "Congratulations" a bit too enthusiastic? Also, I would get rid of "A judge .." in favor of "Your paper has been ...". It doesn't matter to the user that there are judges etc, all that's important is that it has been accepted/rejected/etc

File: indico/modules/events/papers/templates/emails/judgment_to_author.html [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+        <p>A judge has <strong>rejected</strong> your paper <strong>{{ contribution.title }}</strong>.</p>
+        <p>No further actions are required.</p>
+    {% else %}
+        <p>A judge has requested some <strong>changes</strong> to your paper <strong>{{ contribution.title }}</strong>.</p>
+    {% endif %}
+
+{% endblock %}
+
+{% block content_body %}
+    {% if paper.last_revision._judgment_comment %}
```
Comment (ThiefMaster) [inline indico/modules/events/papers/templates/emails/judgment_to_author.html]: Why not add a `judgment_comment` property to the `Paper` class?

File: indico/modules/events/papers/templates/emails/paper_assignment.html [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+        New paper review assignment
+    {%- else -%}
+        Removed from paper review assignment
+    {%- endif %}
+{% endblock %}
+
+{% block content_header %}
+    <p>Dear {{ receiver.first_name }},</p>
+    {% if contribs | length == 1 %}
+        {% if assign %}
```
Comment (ThiefMaster) [inline indico/modules/events/papers/templates/emails/paper_assignment.html]: I'd `{% set contrib = contribs|first %}` and use that instead of `contribs[0]`

File: indico/modules/events/papers/templates/emails/base.html (hunk 1 above)
Comment (OmeGak) [inline indico/modules/events/papers/templates/emails/base.html]: :fire:

File: indico/modules/events/papers/templates/emails/base.html (hunk 2 above)
Comment (OmeGak) [inline indico/modules/events/papers/templates/emails/base.html]: :dark_sunglasses: 
:smiley:

File: indico/modules/events/papers/templates/emails/base.html (hunk 3 above)
Comment (OmeGak) [inline indico/modules/events/papers/templates/emails/base.html]: :sunglasses:

File: indico/modules/events/papers/templates/emails/judgment_reset_to_author.html (hunk 12 above)
Comment (nop33) [inline indico/modules/events/papers/templates/emails/judgment_reset_to_author.html]: nah, I think paper page is fine..

Comment (ThiefMaster) [inline indico/modules/events/papers/templates/emails/judgment_reset_to_author.html]: sounds a bit weird tbh. "To see your paper, use the following link"?

File: indico/modules/events/papers/notifications.py [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+        template = get_template_module('events/papers/emails/review_submission_to_judge.html', event=event,
+                                       review=review, contribution=review.revision.paper.contribution, receiver=judge)
+        email = make_email(to_list=judge.email, template=template, html=True)
+        send_email(email, event=event, module='Papers', user=session.user)
+
+
+def notify_paper_judgment(paper, reset=False):
+    event = paper.event_new
+    receiver = paper.last_revision.submitter
+    if paper_reviewing_settings.get(event, 'notify_author_on_judgment') and not reset:
```
Comment (ThiefMaster) [inline indico/modules/events/papers/notifications.py]: This logic can be simplified:

```python
if reset:
    ...
elif paper_reviewing_settings.get(event, 'notify_author_on_judgment'):
    ...
```

File: indico/modules/events/papers/notifications.py [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+def notify_paper_assignment(user, role, contributions, event, assign):
+    template = get_template_module('events/papers/emails/paper_assignment.html', event=event, contribs=contributions,
+                                   receiver=user, assign=assign, role=role)
+    email = make_email(to_list=user.email, template=template, html=True)
+    send_email(email, event=event, module='Papers', user=session.user)
+
+
+def notify_comment(person, paper, comment, submitter):
+    event = paper.event_new
+    receiver_name = person.first_name or 'user'
```
Comment (ThiefMaster) [inline indico/modules/events/papers/notifications.py]: You already do that in the template so receiver should just be `person`

File: indico/modules/events/papers/operations.py [hunk 18]
```diff
@@ -139,16 +158,17 @@ def create_paper_revision(paper, submitter, files):
                        _contribution=paper.contribution)
         pf.save(f.file)
     db.session.flush()
+    db.session.expire(revision._contribution, ['_paper_last_revision'])
```
Comment (ThiefMaster) [inline indico/modules/events/papers/operations.py]: Is this still needed? IIRC the problem you had there was caused by the memoization of the `paper` property.

Comment (nop33) [inline indico/modules/events/papers/operations.py]: It's needed. Otherwise:
```
File "/home/iliastrichopoulos/dev/indico/indico/modules/events/papers/notifications.py", line 27, in notify_paper_revision_submission
    event = revision.paper.event_new
AttributeError: 'NoneType' object has no attribute 'event_new'
```

File: indico/modules/events/papers/controllers/base.py [hunk 19]
```diff
@@ -95,4 +95,7 @@ def _check_paper_protection(self):
         Override this in case you want to check for more specific
         privileges than the generic "can access".
         """
-        return self.contribution.can_access(session.user)
+        return (self.contribution.is_user_associated(session.user, check_abstract=True) or
```
Comment (ThiefMaster) [inline indico/modules/events/papers/controllers/base.py]: wrong place.  check what inherits from this class... this code should go into the override in RHPaperTimeline

File: indico/modules/events/papers/controllers/display.py [hunk 20]
```diff
@@ -59,6 +59,12 @@ class RHPaperTimeline(RHPaperBase):
     def _process(self):
         return render_paper_page(self.paper, view_class=WPDisplayCallForPapers)
 
+    def _check_paper_protection(self):
+        return (self.contribution.is_user_associated(session.user, check_abstract=True) or
+                self.paper.cfp.is_manager(session.user) or
```
Comment (ThiefMaster) [inline indico/modules/events/papers/controllers/display.py]: I'd use `self.event_new.cfp` simply to make it clear that this is not related to that specific paper
